core/lib/prover_interface/src/inputs.rs

```diff
@@ -160,7 +190,19 @@ impl StoredObject for VMRunWitnessInputData {
         format!("vm_run_data_{key}.bin")
     }
 
-    serialize_using_bincode!();
+    fn serialize(&self) -> Result<Vec<u8>, BoxedError> {
+        bincode::serialize(self).map_err(Into::into)
+    }
+
+    fn deserialize(bytes: Vec<u8>) -> Result<Self, BoxedError> {
+        if let Ok(res) = bincode::deserialize::<VMRunWitnessInputData>(&bytes).map_err(From::from) {
+            return Ok(res);
+        } else {
+            bincode::deserialize::<VMRunWitnessInputDataLegacy>(&bytes)
+                .map_err(From::from)
+                .map(Into::into)
```

Comment:
nit: you can use `or_else` instead of `if let Ok(res) ... return Ok(res);`, also let's use either From::from or Into::into for consistency

also a small comment about why we need Legacy struct is preferred